### PR TITLE
Recurrent PPO

### PR DESCRIPTION
**Pull request by araffin**
## Description

Experimental version of PPO with LSTM policy.

Current status: usable but not polished, see https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/53#issuecomment-1074371013

Missing:
- [x] benchmark https://wandb.ai/sb3/no-vel-envs/reports/PPO-vs-RecurrentPPO-aka-PPO-LSTM-on-environments-with-masked-velocity--VmlldzoxOTI4NjE4 (and models in https://wandb.ai/openrlbenchmark/sb3?workspace=user-araffin )
- [x] more tests (reproducibility + identity learning/callback)
- [x] reduce code duplication in buffer
- [x] document buffer + rest
- [x] more testing of the different architecture (shared vs separate lstm for actor/critic)

Known issue: if the model was train on GPU and tested on CPU, a warning will be issued because it cannot unpickle the lstm initial states. This is ok as they will be reset anyway in `setup_model()` and it doesn't affect prediction.

## Context

- [x] I have raised an issue to propose this change ([required](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/CONTRIBUTING.md))
closes https://github.com/DLR-RM/stable-baselines3/issues/18

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] The functionality/performance matches that of the source (**required** for new training algorithms or training-related features).
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have included an example of using the feature (*required for new features*).
- [x] I have included baseline results (**required** for new training algorithms or training-related features).
- [x] I have updated the documentation accordingly.
- [x] I have updated the changelog accordingly (**required**).
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: sb3_contrib/common/recurrent/policies.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+            features_extractor_class,
+            features_extractor_kwargs,
+            normalize_images,
+            optimizer_class,
+            optimizer_kwargs,
+        )
+
+        self.shared_lstm = shared_lstm
+        self.enable_critic_lstm = enable_critic_lstm
+        self.lstm_actor = nn.LSTM(self.features_dim, lstm_hidden_size, num_layers=n_lstm_layers)
```
Comment (karevn) [inline sb3_contrib/common/recurrent/policies.py:108]: What do you think about adding a dropout option here, or, perhaps adding LSTM kwargs?

**Comment by jjshoots** [comment]
Hi there, I'm not sure where to mention this, but I'm currently using the work from this branch, and model.save() has an issue where the model saves fine, but it never returns. The result is the code just hangs indefinitely. Thought I'd let someone know.

This is my system:
```
>>> sb3.get_system_info()
OS: Linux-5.13.0-30-generic-x86_64-with-glibc2.29 #33~20.04.1-Ubuntu SMP Mon Feb 7 14:25:10 UTC 2022
Python: 3.8.10
Stable-Baselines3: 1.4.0
PyTorch: 1.10.0+cu102
GPU Enabled: True
Numpy: 1.22.2
Gym: 0.22.0
```

**Comment by araffin** [comment]
> Hi there, I'm not sure where to mention this, but I'm currently using the work from this branch, and model.save() has an issue where the model saves fine, but it never returns. The result is the code just hangs indefinitely. Thought I'd let someone know.
> 
> This is my system:
> 
> ```
> >>> sb3.get_system_info()
> OS: Linux-5.13.0-30-generic-x86_64-with-glibc2.29 #33~20.04.1-Ubuntu SMP Mon Feb 7 14:25:10 UTC 2022
> Python: 3.8.10
> Stable-Baselines3: 1.4.0
> PyTorch: 1.10.0+cu102
> GPU Enabled: True
> Numpy: 1.22.2
> Gym: 0.22.0
> ```

Hello,
I'm back from a 4 week break... could you please provide a minimal code to reproduce the issue? (without a custom env)

**Comment by jjshoots** [comment]
> > Hi there, I'm not sure where to mention this, but I'm currently using the work from this branch, and model.save() has an issue where the model saves fine, but it never returns. The result is the code just hangs indefinitely. Thought I'd let someone know.
> > This is my system:
> > ```
> > >>> sb3.get_system_info()
> > OS: Linux-5.13.0-30-generic-x86_64-with-glibc2.29 #33~20.04.1-Ubuntu SMP Mon Feb 7 14:25:10 UTC 2022
> > Python: 3.8.10
> > Stable-Baselines3: 1.4.0
> > PyTorch: 1.10.0+cu102
> > GPU Enabled: True
> > Numpy: 1.22.2
> > Gym: 0.22.0
> > ```
> 
> Hello, I'm back from a 4 week break... could you please provide a minimal code to reproduce the issue? (without a custom env)
> 
> Note: current version only supports gym 0.19 (gym 0.21 will be supported soon (already on master), 0.22 introduces new breaking changes...)

Apologies, I think it may be an issue on my end regarding the env I'm using. I'm currently using Pistonball of  PettingZoo using [this hack](https://towardsdatascience.com/multi-agent-deep-reinforcement-learning-in-15-lines-of-code-using-pettingzoo-e0b963c0820b). I can't get it to replicate using normal gym, even gym 0.22.

**Comment by araffin** [comment]
@rvdweerd 

Replying to your email here so it serves others:
 > Is it possible for you to comment on the status of the PPO+LSTM implementation?
>    I gather it is still in development and not working yet, any chance of a testable version in the near future?

The implementation is working (it was tested on different env, including cartpole without velocity which can be solved only with RNN) but it needs some polishing and additional testing.

>    Would it also be possible to use action-maskable PPO policies + LSTM?

In theory yes, in practice, I will keep the PPO LSTM implementation as simple as possible, but nothing prevents you from combining the Maskable PPO with it ;)

>   Would the implementation include Dict observation spaces?

Yes, I think that's one of the thing that need to be done before merging. Contributions are welcomed ;)

And anyway, I would always recommend to start with frame-stacking and feedforward network as a baseline for POMDP, it is usually competitive with LSTM.

**Comment by Miffyli** [comment]
@rrfaria This was fixed in a recent PR in SB3 [here](https://github.com/DLR-RM/stable-baselines3/pull/839). Try doing [these](https://github.com/DLR-RM/stable-baselines3/pull/839/files#diff-3b26bc73c977b035e2ed9bd03968728eb84491209749a38ff08cb31da9b0400eL316-L322) changes manually and see if that helps.

**Comment by JonathanvanWestendorp** [comment]
Hi! When using multiple environments, it does not seem to be running on multiple cores of the CPU. Resulting in very slow training. Exactly the same setup works fine on the normal PPO implementation. Is this something that still has to be addressed?

**Comment by araffin** [comment]
> it does not seem to be running on multiple cores of the CPU. Resulting in very slow training.

could you provide a minimal example?
multiple cores for collecting data or doing gradient update?

> Resulting in very slow training.

this highly depends on your hyperparameters, especially the `n_steps`, LSTM PPO is slower anyway compared to the feedforward one.

**Comment by JonathanvanWestendorp** [comment]
For now we have:

model = Model("MlpLstmPolicy",
              env=dummy_env,
              n_steps=128,
              batch_size=14,
              gamma=0.5,
              n_epochs=5,
              ent_coef=0.0,
              policy_kwargs=policy_kwargs,
              verbose=1,
              tensorboard_log=tensorboard_log,
              device="auto")

where Model is the RecurrentPPO. While training htop shows that only 1 core is in use except prior to logging the new stats, then it suddenly uses all the cores. With stats i mean:
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 252         |
|    ep_rew_mean          | 12.4        |
| time/                   |             |
|    fps                  | 451         |
|    iterations           | 5           |
|    time_elapsed         | 45          |
|    total_timesteps      | 20480       |
| train/                  |             |
|    approx_kl            | 0.021662503 |
|    clip_fraction        | 0.297       |
|    clip_range           | 0.2         |
|    entropy_loss         | -0.907      |
|    explained_variance   | 0           |
|    learning_rate        | 0.0003      |
|    loss                 | 0.336       |
|    n_updates            | 20          |
|    policy_gradient_loss | -0.0208     |
|    value_loss           | 0.56        |
-----------------------------------------

We have 32 threads, so our vectorized environment exists of 32 envs.
env = make_vec_env(StockEnv, env_kwargs={**env_kwargs, 'df': df}, n_envs=args.n_envs, vec_env_cls=SubprocVecEnv)

**Comment by JonathanvanWestendorp** [comment]
> > it does not seem to be running on multiple cores of the CPU. Resulting in very slow training.
> 
> could you provide a minimal example? multiple cores for collecting data or doing gradient update?
> 
> > Resulting in very slow training.
> 
> this highly depends on your hyperparameters, especially the `n_steps`, LSTM PPO is slower anyway compared to the feedforward one.

For now we have:

model = Model("MlpLstmPolicy",
              env=dummy_env,
              n_steps=128,
              batch_size=14,
              gamma=0.5,
              n_epochs=5,
              ent_coef=0.0,
              policy_kwargs=policy_kwargs,
              verbose=1,
              tensorboard_log=tensorboard_log,
              device="auto")

where Model is the RecurrentPPO. While training htop shows that only 1 core is in use except prior to logging the new stats, then it suddenly uses all the cores for one sec and goes back to just using 1 afterwards. With stats I mean the little table showing the rollout/ time/ train/ etc in the terminal.

We have 32 threads, so our vectorized environment exists of 32 envs.
env = make_vec_env(Env, env_kwargs={**env_kwargs, 'df': df}, n_envs=args.n_envs, vec_env_cls=SubprocVecEnv)

When using the normal PPO we see usage of all cores in htop. Only when our evaluation callback is called you see a dip to one core because of evaluation. We use the same training loop for PPO and RecurrentPPO. We just initialize, set env as above. and start learning.

**Comment by andrewwarrington** [comment]
Hi folks,  I am looking at using this code, and just confirming a particular of the internals:

When doing the backprop through the action probabilities ([here](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/feat/ppo-lstm/sb3_contrib/ppo_recurrent/ppo_recurrent.py#L381)), you just take a single step in the RNN, i.e. the gradient with respect to the RNN parameters given just the previous RNN state and input?  

The alternative to this is re-processing a short sequence and backpropping over that (truncated back-prop through time (t-BPTT)).  This will give different gradients on the weights.  This would be analogous to `RECURRENT_SEQ_LEN` in [this implementation](https://gitlab.com/ngoodger/ppo_lstm/-/blob/master/recurrent_ppo.ipynb).  

I could believe that this wouldn't be too difficult to add to the code.  The buffer just generates a sequence of length `RECURRENT_SEQ_LEN`, instead of calling `self.policy.evaluate_actions` directly, it calls out to a short subroutine for re-processing the sequence, and still returns `values, log_prob, entropy`.  Would be happy to contribute to this effort.  

More generally, is there a particular consensus on whether (t-)BPTT is required to learn good policies?  

Thanks a ton for grafting hard on this recurrent PPO!
A

File: sb3_contrib/common/recurrent/policies.py (hunk 1 above)
Comment (araffin) [reply sb3_contrib/common/recurrent/policies.py:108]: why not, might be helpful

Comment (araffin) [reply sb3_contrib/common/recurrent/policies.py]: done ;)

**Comment by araffin** [comment]
@andrewwarrington 

> When doing the backprop through the action probabilities ([here](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/feat/ppo-lstm/sb3_contrib/ppo_recurrent/ppo_recurrent.py#L381)), you just take a single step in the RNN, i.e. the gradient with respect to the RNN parameters given just the previous RNN state and input?

Actually not, see the shape in the buffer https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/feat/ppo-lstm/sb3_contrib/common/recurrent/buffers.py#L164
I just pushed proper masking for the padded sequences. We do backprop for the sequences collected during the `n_steps` of the rollout.

**Comment by HamiltonWang** [comment]
for the time being, is there any way we can feedback the older data back to the trainer as input to mimic a crude version of LSTM? any sample code to do that?

**Comment by araffin** [comment]
> for the time being, is there any way we can feedback the older data back to the trainer as input to mimic a crude version of LSTM? any sample code to do that?

Hello,
you can already use this PR if you need to use recurrent PPO (see install from source in our doc), otherwise you can use frame stacking or history wrapper (see code in the RL Zoo).

**Review by vineetvermait — APPROVED** [review]

**Comment by henrydeclety** [comment]
when LSTM for A2C?

**Comment by araffin** [comment]
> when LSTM for A2C?

a2c is a special case of ppo ;) (cc @vwxyzjn )

**Comment by vwxyzjn** [comment]
@henrydeclety see https://github.com/vwxyzjn/a2c_is_a_special_case_of_ppo. We have a paper coming out soon...

**Comment by vwxyzjn** [comment]
The preprint of the paper is out at https://arxiv.org/abs/2205.09123 @henrydeclety :)

**Comment by HamiltonWang** [comment]
> Hello, you can already use this PR if you need to use recurrent PPO (see install from source in our doc), otherwise you can use frame stacking or history wrapper (see code in the RL Zoo).

I’ll give it a try

**Comment by EloyAnguiano** [comment]
How could I configure the maximum sequence length for the LSTM?

**Comment by philippkiesling** [comment]
@EloyAnguiano As far as I could tell from the code, the implementation in SB3 does not have a sequence length, but saves the hidden state between steps of your environment and then uses it as input. So the maximum sequence length for the lstm would be the number of steps (n_steps) before you update your policy.
